Neos.EventSourcedNeosAdjustments/Classes/Domain/Service/NodeShortcutResolver.php

```diff
@@ -12,89 +12,127 @@
-use Neos\Neos\Controller\Exception\NodeNotFoundException;
-use Neos\Neos\Service\LinkingService;
+use Neos\EventSourcedNeosAdjustments\EventSourcedRouting\Exception\InvalidShortcutException;
+use Neos\EventSourcedNeosAdjustments\EventSourcedRouting\Projection\DocumentUriPathFinder;
+use Neos\EventSourcedNeosAdjustments\EventSourcedRouting\ValueObject\DocumentNodeInfo;
 use Neos\Flow\Annotations as Flow;
+use Neos\Flow\ResourceManagement\ResourceManager;
+use Neos\Media\Domain\Repository\AssetRepository;
+use Neos\Neos\Controller\Exception\NodeNotFoundException;
+use Psr\Http\Message\UriInterface;
 
 /**
  * Can resolve the target for a given shortcut.
+ * Used for Neos Routing ({@see EventSourcedFrontendNodeRoutePartHandler}), and redirects to a shortcut target when visiting the shortcut itself.
  *
  * @Flow\Scope("singleton")
  */
 class NodeShortcutResolver
 {
-    /**
-     * @Flow\Inject
-     * @var LinkingService
-     */
-    protected $linkingService;
+    private DocumentUriPathFinder $documentUriPathFinder;
 
-    /**
-     * @Flow\Inject
-     * @var NodeTypeConstraintFactory
-     */
-    protected $nodeTypeConstraintFactory;
+    private AssetRepository $assetRepository;
+
+    private ResourceManager $resourceManager;
+
+    public function __construct(DocumentUriPathFinder $documentUriPathFinder, AssetRepository $assetRepository, ResourceManager $resourceManager)
+    {
+        $this->documentUriPathFinder = $documentUriPathFinder;
+        $this->assetRepository = $assetRepository;
+        $this->resourceManager = $resourceManager;
+    }
 
     /**
-     * Resolves a shortcut node to the target. The return value can be
-     *
-     * * string (in case the target is a plain text URI, a node:// URI, an asset:// URI or a node)
-     * * NULL in case the shortcut cannot be resolved
+     * "adapter" for {@see resolveNode} when working with NodeAddresses.
      *
-     * @param ContentSubgraphInterface $subgraph
-     * @param NodeInterface $shortcut
      * @param NodeAddress $nodeAddress
-     * @param UriBuilder $uriBuilder
-     * @param string $format
-     * @return string|null
+     * @return NodeAddress|UriInterface NodeAddress is returned if we want to link to another node (i.e. node is NOT a shortcut node; or target is a node); or UriInterface for links to fixed URLs (Asset URLs or external URLs)
+     * @throws InvalidShortcutException
      * @throws NodeNotFoundException
-     * @throws \Neos\Flow\Mvc\Routing\Exception\MissingActionNameException
      */
-    public function resolveShortcutTarget(ContentSubgraphInterface $subgraph, NodeInterface $shortcut, NodeAddress $nodeAddress, UriBuilder $uriBuilder, string $format): ?string
+    public function resolveShortcutTarget(NodeAddress $nodeAddress)
+    {
+        $node = $this->documentUriPathFinder->getOneByIdAndDimensionSpacePointHash($nodeAddress->getNodeAggregateIdentifier(), $nodeAddress->getDimensionSpacePoint()->getHash());
```

Comment:
comment: "we do not need to pass in the ContentStreamIdentifier, because we are always dealing with the live workspace"

Reply:
Variable name: $documentNodeInfo and not "$node" (with $node, I think of TraversableNode)